component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,220 @@
+ * canHandle method will confirm whether this request can be handled by this authenticator or not.
+ */
+public class DPoPAuthenticationHandler extends AuthenticationHandler {
+
+    private static final Log log = LogFactory.getLog(DPoPAuthenticationHandler.class);
+
+    @Override
+    protected AuthenticationResult doAuthenticate(MessageContext messageContext) {
+
+        AuthenticationResult authenticationResult = new AuthenticationResult(AuthenticationStatus.FAILED);
+        AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+        AuthenticationRequest authenticationRequest = authenticationContext.getAuthenticationRequest();
+        if (authenticationRequest != null) {
+
+            String authorizationHeader = authenticationRequest.getHeader(HttpHeaders.AUTHORIZATION);
+            if (StringUtils.isNotBlank(authorizationHeader) &&
+                    authorizationHeader.startsWith(DPoPConstants.OAUTH_DPOP_HEADER)) {
+                String accessToken = null;
+                String[] bearerToken = authorizationHeader.split(" ");
+                if (bearerToken.length == 2) {
+                    accessToken = bearerToken[1];
+                }
+
+                OAuth2TokenValidationService oAuth2TokenValidationService = new OAuth2TokenValidationService();
+                OAuth2TokenValidationRequestDTO requestDTO = new OAuth2TokenValidationRequestDTO();
+                OAuth2TokenValidationRequestDTO.OAuth2AccessToken token = requestDTO.new OAuth2AccessToken();
+                token.setIdentifier(accessToken);
+                token.setTokenType(DPoPConstants.OAUTH_HEADER);
+                requestDTO.setAccessToken(token);
+
+                //TODO: If these values are not set, validation will fail giving an NPE. Need to see why that happens
+                OAuth2TokenValidationRequestDTO.TokenValidationContextParam contextParam = requestDTO.new
+                        TokenValidationContextParam();
+                contextParam.setKey("dummy");
+                contextParam.setValue("dummy");
+
+                OAuth2TokenValidationRequestDTO.TokenValidationContextParam[] contextParams = {contextParam};
+                requestDTO.setContext(contextParams);
+
+                OAuth2ClientApplicationDTO clientApplicationDTO = oAuth2TokenValidationService
+                        .findOAuthConsumerIfTokenIsValid
+                                (requestDTO);
+                OAuth2TokenValidationResponseDTO responseDTO = clientApplicationDTO.getAccessTokenValidationResponse();
+
+                getAuthenticationResult(authenticationResult, responseDTO, authorizationHeader,
+                        authenticationRequest);
+
+                String consumerKey = clientApplicationDTO.getConsumerKey();
+
+                authenticationResult.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+
+                if (StringUtils.isNotEmpty(responseDTO.getAuthorizedUser())) {
```

Comment:
This part also you can move into a new private method.

Reply:
Fixed with [a4c65c3](https://github.com/kalanika/identity-oauth-addons/commit/a4c65c37b8b7eb980f5cb1b9569a27044541faed).